src/py/flwr/simulation/ray_transport/ray_actor.py

```diff
@@ -0,0 +1,388 @@
+# Copyright 2023 Flower Labs. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Ray-based Flower Actor and ActorPool implementation."""
+
+
+import threading
+import traceback
+from abc import ABC
+from logging import ERROR, WARNING
+from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Type, Union
+
+import ray
+from ray import ObjectRef
+from ray.util.actor_pool import ActorPool
+
+from flwr import common
+from flwr.common.logger import log
+
+# All possible returns by a client
+ClientRes = Union[
+    common.GetPropertiesRes, common.GetParametersRes, common.FitRes, common.EvaluateRes
+]
+
+
+class ClientException(Exception):
+    """Raised when client side logic crashes with an exception."""
+
+    def __init__(self, message: str):
+        div = ">" * 7
+        self.message = "\n" + div + "A ClientException occurred." + message
+        super().__init__(self.message)
+
+
+class VirtualClientEngineActor(ABC):
+    """Abstract base class for VirtualClientEngine Actors."""
+
+    def terminate(self) -> None:
+        """Manually terminate Actor object."""
+        log(WARNING, "Manually terminating %s}", self.__class__.__name__)
+        ray.actor.exit_actor()
+
+    def run(self, job_fn: Callable[[], ClientRes], cid: str) -> Tuple[str, ClientRes]:
+        """Run a client workload."""
+        # Execute tasks and return result
+        # return also cid which is needed to ensure results
+        # from the pool are correctly assigned to each ClientProxy
+        try:
+            job_results = job_fn()
+        except Exception as ex:
+            client_trace = traceback.format_exc()
+            message = (
+                "\n\tSomething went wrong when running your client workload."
+                "\n\tClient "
+                + cid
+                + " crashed when the "
+                + self.__class__.__name__
+                + " was running its workload."
+                "\n\tException triggered on the client side: " + client_trace,
+            )
+            raise ClientException(str(message)) from ex
+
+        return cid, job_results
+
+
+@ray.remote
+class DefaultActor(VirtualClientEngineActor):
+    """A Ray Actor class that runs client workloads."""
```

Comment:
Do we need this, or could we just annotate `VirtualClientEngineActor` with `@ray.remote`, and not make it an `ABC`?

Reply:
It's not possible to inherit from classes with @ray.remote decorator.